### PR TITLE
Dockerize ClamAV and Sidekiq in local development

### PR DESCRIPTION
**Pull request by f1337**
- Extracts ClamAV and Sidekiq processes into their own Docker containers in preperation for moving to Kubernetes.
- Logs ClamAV results via Rails.logger (for Loki).
- ClamAV config is now copied to /etc (default config location) instead of passed as an argument on each run.
- Foreman is not completely removed here as it's still used in higher environments.

### CODE REVIEW

File: lib/shrine/plugins/validate_virus_free.rb [hunk 1]
```diff
@@ -9,7 +9,8 @@ module AttacherMethods
         def validate_virus_free(message: nil)
           cached_path = get.download.path
           result = Common::VirusScan.scan(cached_path)
-          # TODO: Log a the full result to sentry
+          # TODO: Log a the full result to ~sentry~ Loki
+          # Logger.info('Common::VirusScan result', result)
```
Comment (f1337) [inline lib/shrine/plugins/validate_virus_free.rb:13]: I'd love to go ahead and fix this TODO if we can. What's it take to log to Loki?

File: lib/common/virus_scan.rb [hunk 2]
```diff
@@ -7,8 +7,7 @@ module VirusScan
     def scan(file_path)
       # `clamd` runs within service group, needs group read
       File.chmod(0o640, file_path)
-      args = ['--config', Rails.root.join('config', 'clamd.conf').to_s]
-      ClamScan::Client.scan(location: file_path, custom_args: args)
+      ClamScan::Client.scan(location: file_path)
```
Comment (f1337) [inline lib/common/virus_scan.rb:10]: Moved the custom config to the Dockerfile, because that's where it truly belongs.

File: docker-compose.yml [hunk 3]
```diff
@@ -42,15 +46,16 @@ services:
       "Settings.saml.key_path": "/srv/vets-api/src/config/certs/vetsgov-localhost.key"
       "Settings.redis.app_data.url": "redis://redis:6379"
       "Settings.redis.sidekiq.url": "redis://redis:6379"
-      "Settings.binaries.clamdscan": "clamscan" # Not running a separate process within the container for clamdscan, so we use clamscan which requires no daemon
```
Comment (f1337) [inline docker-compose.yml:45]: We will probably want to restore the ability for local dev to skip clamav. Then again, it's really fast, so :shrug:.

File: docker-compose.yml [hunk 4]
```diff
@@ -1,5 +1,9 @@
 version: '3.4'
 services:
+  clamav:
+    image: mkodockx/docker-clamav
+    ports:
+      - "3310:3310"
```
Comment (f1337) [inline docker-compose.yml:6]: The magic is mostly here. 🪄

File: config/initializers/clamscan.rb [hunk 5]
```diff
@@ -2,4 +2,5 @@
 
 ClamScan.configure do |clam_config|
   clam_config.client_location = Settings.binaries.clamdscan
+  clam_config.default_scan_options = { stdout: false }
```
Comment (f1337) [inline config/initializers/clamscan.rb:5]: I'm unsure if this is really necessary. But stdout pollution is evil. So I suppressed it.

File: config/clamd.conf [hunk 6]
```diff
@@ -1,5 +1,3 @@
 Foreground yes
-DatabaseDirectory /srv/vets-api/clamav/database
-LocalSocket /srv/vets-api/clamav/clamd.ctl
 TCPSocket 3310
-TCPAddr 127.0.0.1
\ No newline at end of file
+TCPAddr clamav
```
Comment (f1337) [inline config/clamd.conf:3]: The hostname of the container. If we name the pods the same, this might work unchanged in k8s?

File: Procfile [hunk 7]
```diff
@@ -1,4 +1,2 @@
 web: bundle exec puma -p 3000 -C ./config/puma.rb
 job: bundle exec sidekiq -q critical,4 -q tasker,3 -q default,2 -q low,1
-freshclam: /usr/bin/freshclam -d --config-file=config/freshclam.conf
-clamd: /usr/sbin/clamd -c config/clamd.conf
```
Comment (f1337) [inline Procfile:4]: 2 down, 1 to go (job).

File: Dockerfile [hunk 8]
```diff
@@ -12,13 +12,14 @@ SHELL ["/bin/bash", "-c"]
 RUN groupadd -g $userid -r vets-api && \
     useradd -u $userid -r -m -d /srv/vets-api -g vets-api vets-api
 RUN apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install -y \
-    dumb-init clamav clamdscan clamav-daemon imagemagick pdftk curl poppler-utils libpq5 vim
+    dumb-init clamdscan imagemagick pdftk curl poppler-utils libpq5 vim
```
Comment (f1337) [inline Dockerfile:15]: Don't need clam or the daemon, just the client.

File: Dockerfile [hunk 9]
```diff
@@ -12,13 +12,14 @@ SHELL ["/bin/bash", "-c"]
...
 # The pki work below is for parity with the non-docker BRD deploys to mount certs into
 # the container, we need to get rid of it and refactor the configuration bits into
 # something more continer friendly in a later bunch of work
 RUN mkdir -p /srv/vets-api/{clamav/database,pki/tls,secure,src} && \
     chown -R vets-api:vets-api /srv/vets-api && \
     ln -s /srv/vets-api/pki /etc/pki
+COPY config/clamd.conf /etc/clamav/clamd.conf
```
Comment (f1337) [inline Dockerfile:22]: because overriding the AV config at runtime feels wrong.

File: docker-compose.yml [hunk 10]
```diff
@@ -42,17 +47,51 @@ services:
...
       POSTGRES_PASSWORD: "${POSTGRES_PASSWORD:-password}"
     depends_on:
+      - clamav
+      - postgres
+      - redis
+    links:
+      - clamav
+      - postgres
+      - redis
+  worker:
```
Comment (f1337) [inline docker-compose.yml:62]: Sidekiq worker extracted too 🤘🏻 .

File: lib/shrine/plugins/validate_virus_free.rb (hunk 1 above)
Comment (dginther) [reply lib/shrine/plugins/validate_virus_free.rb:13]: Set the docker logging driver to loki and point it at the loki server. an example: https://github.com/department-of-veterans-affairs/devops/blob/a04b11a7208379360672ae4dbb24049affd1df1b/ansible/deployment/config/vets-api-server-vagov-dev.yml#L78

File: docker-compose.yml (hunk 4 above)
Comment (dginther) [reply docker-compose.yml:6]: likely we do not want to pull this from Dockerhub, as we may run into quota issues. Recommend setting up a repo that mirrors the clamav container to ECR, similar to the way we do here: https://github.com/department-of-veterans-affairs/vsp-infra-sentry/blob/main/.github/workflows/mirror-images.yaml and then using the container from there.

File: docker-compose.yml [hunk 11]
```diff
@@ -1,5 +1,9 @@
 version: '3.4'
 services:
+  clamav:
+    image: clamav/clamav
+    ports:
+      - "3310:3310"
   redis:
     image: redis:5.0-alpine
```
Comment (dginther) [inline docker-compose.yml:8]: Same comment as about the clamav container.

File: docker-compose.yml (hunk 4 above)
Comment (f1337) [reply docker-compose.yml:6]: right on, so is the idea that i would create such a yaml in vets-api/.github/workflows/mirror-images.yml?

Comment (f1337) [reply docker-compose.yml:6]: curious: is that an ansible file? or is that argo?

Comment (dginther) [reply docker-compose.yml:6]: that is a github actions workflow

File: Makefile [hunk 12]
```diff
@@ -148,5 +148,5 @@ else
 endif
 
 .PHONY: up
-up: db  ## Starts the server and associated services with docker-compose, use `clam=1 make up` to run ClamAV
-	@$(BASH_DEV) "rm -f tmp/pids/server.pid && foreman start -m ${FOREMAN_ARG}"
+up:
```
Comment (rileyanderson) [inline Makefile:151]: Do we need to keep `db` here to set up the databasae?

Comment (f1337) [reply Makefile:151]: Doh! Good catch!

Comment (f1337) [reply Makefile]: fixed.

File: lib/shrine/plugins/validate_virus_free.rb (hunk 1 above)
Comment (f1337) [reply lib/shrine/plugins/validate_virus_free.rb]: i think since Loki is live now, as long as I use `Rails.logger`, this should go to Loki. i hope? 🤞🏻

**Review by ericboehs — APPROVED** [review]

**Pull request merged**